pennylane_sf/gbs.py

```diff
@@ -124,46 +134,110 @@ def execute(self, queue, observables, parameters=None, **kwargs):
+        # There is no engine to reset in this case
+        if not self.analytic and self.use_cache and self.samples is not None:
+            return
+        super().reset()
+
     def pre_measure(self):
         self.eng = sf.Engine("gaussian", backend_options={"cutoff_dim": self.cutoff})
 
         if self.analytic:
             results = self.eng.run(self.prog)
+        elif self.use_cache and self.samples is not None:  # use the cached samples
+            return
         else:
             results = self.eng.run(self.prog, shots=self.shots)
 
         self.state = results.state
         self.samples = results.samples
 
+    def _reparametrize_probability(self, p):
+        """Takes an input probability distribution of :math:`A` and rescales it to the probability
+        distribution of :math:`WAW`.
+
+        Args:
+            p (array): the probability distribution of :math:`A`
+
+        Returns:
+            array: the probability distribution of :math:`WAW`
+        """
+        Z_inv = self._calculate_z_inv(self._WAW)
+        ind_all_wires = np.ndindex(*[self.cutoff] * self.num_wires)
+
+        for s in ind_all_wires:
+            res = np.prod(np.power(self._params, s))
+            p[tuple(s)] *= res * Z_inv / self.Z_inv
+
+        return p
+
+    def _marginal_over_wires(self, wires, array):
+        """Calculates the marginal distribution over the specified wires by summing over the
+        remaining.
+
+        The flat input array is reshaped to have ``self.num_wires + 1`` axes.
+
+        Args:
+            wires (int or Iterable[Number, str]]): the wires to find marginals over
+            array (array): the input array
+
+        Returns:
+            array: the traced-over array
+        """
+        trailing = int(array.size / self.cutoff ** self.num_wires)
+
+        array = array.reshape([self.cutoff] * self.num_wires + [trailing])
+
+        all_indices = set(range(self.num_wires))
+        requested_indices = set(self.wires.indices(wires))
+
+        trace_over_indices = all_indices - requested_indices  # Find indices to trace over
+        return np.sum(array, axis=tuple(trace_over_indices)).ravel()
+
     def probability(self, wires=None):
         wires = wires or self.wires
         wires = Wires(wires)
 
         if self.analytic:
             return super().probability(wires=wires)
 
-        N = len(wires)
         samples = np.take(self.samples, self.wires.indices(wires), axis=1)
 
-        probs = all_fock_probs_pnr(samples)
-        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+        fock_probs = all_fock_probs_pnr(samples)
+        cutoff = fock_probs.shape[0]
+        diff = max(self.cutoff - cutoff, 0)
+        probs = np.pad(fock_probs, [(0, diff)] * len(wires))
+
+        if self.use_cache:
+            if len(wires) < self.num_wires:
+                raise ValueError("Caching is only supported when returning the probabilities on "
+                                 "all of the wires")
```

Comment:
This could feasibly be fixed to work with subsets of wires, but I realised having a go today that it's nontrivial and would take some work (especially adding tests).

Since I view this as more of an edge case anyway, I'd rather we simply raise an error here and see if we get requests from users for caching + subsets of wires.